clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ConsumerRebalanceListenerCallbackNeededEvent.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals.events;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.internals.ConsumerRebalanceListenerMethodName;
+import org.apache.kafka.common.TopicPartition;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.SortedSet;
+
+/**
+ * Event that signifies that the network I/O thread wants to invoke one of the callback methods on the
+ * {@link ConsumerRebalanceListener}. This event will be processed by the application thread when the next
+ * {@link Consumer#poll(Duration)} call is performed by the user. When processed, the application thread should
+ * invoke the appropriate callback method (based on {@link #methodName()}) with the given partitions.
+ */
+public class ConsumerRebalanceListenerCallbackNeededEvent extends BackgroundEvent {
+
+    private final ConsumerRebalanceListenerMethodName methodName;
+    private final SortedSet<TopicPartition> partitions;
+
+    public ConsumerRebalanceListenerCallbackNeededEvent(ConsumerRebalanceListenerMethodName methodName,
```

Comment:
There is actually a desire to have the ability to execute multiple callbacks (e.g. onAssigned, onRevoked) with a single event. Would it be possible to handle this here? @lianetm Could provide more details.

Reply:
Yes. As comment above, multiple callbacks can happen at the same time if triggered from reconciliation and member leaves/fenced. I would say that we only leave the restriction regarding callback execution on the caller (reconciliation process will trigger them one at a time).

Reply:
@kirktrue How do we go about this one?

Reply:
At this point, the requests to execute the callbacks are enqueued by the membership manager whenever it sees fit to do so. In the application thread, it will pull all the events off the queue and process them by invoking the callbacks. Does there need to be "smarts" in the consumer to manage the ordering of the callback requests?

Reply:
Let me try to re-explain. At the moment, the state machine sends an event for the `onRevoked` callback, waits for its completion, sends an event for the `onAssigned` callback, and waits for its completion. @lianetm and I thought that we should change this to actually trigger them together with a single event in order to reduce the dependency on the foreground thread. This is not done yet but @lianetm will eventually get to this.

With this in mind, I was wondering if we could have already implemented the `ConsumerRebalanceListenerCallbackNeededEvent` in a way which will support this. Otherwise, she will have to refactor it later on. Basically, `ConsumerRebalanceListenerCallbackNeededEvent` would need to support calling both `onRevoke` and `onAssigned`, still one after another in the foreground thread but within a single event.

@lianetm Would it help you if we do it here or are you fine with refactoring this later on?

Reply:
This `callbackNeeded` event will most probably include a list of callbacks rather than a single one, that we'll need to exec sequentially. That intention is basically to improve how callbacks are triggered from the reconciliation, as @dajac described. But I'm ok with refactoring the event and how it's processed later (it's needed for the task [KAFKA-15835](https://issues.apache.org/jira/browse/KAFKA-15835) I have for changing how callbacks are triggered when reconciling, so I'll do it there if we don't have it yet)

Reply:
I approached this change with the idea of making the event hold a list of invocations such that the `ApplicationEventProcessor` could simply loop over them to invoke callbacks serially. But looking more closely at [`ConsumerCoordinator`](https://github.com/apache/kafka/blob/2b99d0e45027c88ae2347fa8f7d1ff4b2b919089/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L397-L420) shows that there are some steps that need to be executed between those two invocations:

1. Invoke `ConsumerRebalanceListener.onPartitionsRevoked()`
2. Invoke `SubscriptionState.subscribeFromPattern()` if there are patterns involved (which we're totally ignoring at present)
3. Invoke `ConsumerPartitionAssignor.onAssignment()`
4. Update the auto-commit timer (which isn't presently available in the application thread)
5. Invoke `SubscriptionState.assignFromSubscribed()`
6. Invoke `ConsumerRebalanceListener.onPartitionsAssigned()`

If we were to perform both invocations in the application thread, we would need to execute steps 1-6 during the processing of the `ConsumerRebalanceListenerCallbackNeededEvent`. It's not necessarily that much code, but there are two main downsides:

1. The clear separation of flow, logic, and data used by the reconciliation logic would no longer reside just in `MembershipManagerImpl`, but would spill over into `ApplicationEventProcessor`
2. The current event is generalized to support execution of _any_ of the three `ConsumerRebalanceListener` methods. We would need to replace that with the more specific processing listed in the six steps above. We would then need to add at least one additional event/processor method to handle the `onPartitionsLost` case.

In the end "it's just code," so we can change it however we want. Given the deadline, I would recommend mulling this over after we have everything functionally correct.

Reply:
Sounds good. Let's change this when @lianetm gets to implementing KAFKA-15835.